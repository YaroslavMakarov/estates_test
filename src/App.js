import React, { useState, useEffect } from "react";
import 'bootstrap/dist/css/bootstrap.min.css';
import { Container, Row, Col } from "react-bootstrap";
import { getEstates } from "./helpers/api";
import Estates from "./components/Estates/Estates"
import './App.css';

const App = () => {
  const [estates, setEstates] = useState([]);

  useEffect(() => {
    getEstates().then(resolve => setEstates(resolve));
  }, []);

  console.log(estates);
  return (
    <Container className="d-flex justify-content-center">
      <Estates estates={estates} />
    </Container>
  );
}

export default App;
