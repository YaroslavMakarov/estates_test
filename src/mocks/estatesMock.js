export const estatesMock = {
    estates: {
        estates: [
            {
                'id': 1,
                'title': 'Shevchenko estate',
                'price': 30000,
                'address': {
                    'street': 'Melnychna',
                    'number': 12
                },
                'preview_img': 'https://cp12.nevsepic.com.ua/62/1353962605-0435309-www.nevsepic.com.ua.jpg',
                'seller': 'Chack Norris',
                'description': 'The best from Chuck',
                'detailed_photos': [
                    'https://www.wtoc.com/resizer/ckPwhlUdYaFcLiBWMX0J033a0Ow=/1200x0/arc-anglerfish-arc2-prod-raycom.s3.amazonaws.com/public/JGJB35IW4JAMROXX3D3MMEF2SU.jpg',
                    'https://img3.goodfon.ru/wallpaper/nbig/c/f7/chuck-norris-chak-norris-legenda-aktior-vzgliad.jpg',
                    'https://cdn.flickeringmyth.com/wp-content/uploads/2020/03/chuck-norris.jpg'
                ]
            },
            {
                "id": 2,
                "title": "Yanukovich estate",
                "price": 3000000,
                "address": {
                    "street": "Prisident",
                    "number": 4
                },
                "preview_img": "https://biz.liga.net/images/general/2018/11/15/thumbnail-tw-20181115111659-8474.jpg?v=1542279867",
                "seller": "Chack Norris",
                "description": "The best from Chuck",
                "detailed_photos": [
                    "https://www.wtoc.com/resizer/ckPwhlUdYaFcLiBWMX0J033a0Ow=/1200x0/arc-anglerfish-arc2-prod-raycom.s3.amazonaws.com/public/JGJB35IW4JAMROXX3D3MMEF2SU.jpg",
                    "https://img3.goodfon.ru/wallpaper/nbig/c/f7/chuck-norris-chak-norris-legenda-aktior-vzgliad.jpg",
                    "https://cdn.flickeringmyth.com/wp-content/uploads/2020/03/chuck-norris.jpg"
                    ]
             },
        ],
    },
};