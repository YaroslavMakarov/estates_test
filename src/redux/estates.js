import { put, takeEvery,
         call } from 'redux-saga/effects';

import { getEstates } from '../helpers/api';       

//action types
const FETCH_ESTATES = "estates_test/estates/fetch";
const SET_ESTATES = "estates_test/estates/fetch";

//action creators
export const fetchPosts = () => ({
    type: FETCH_ESTATES,
});
const setEstates = (estates) => ({
    type: SET_ESTATES,
    estates,
});

//sagas
function* dispatchEstates() {
    const estates = yield call(getEstates);

    yield put(setEstates(estates));
};

export function* dispatchEstatesWatcher() {

    yield takeEvery("estates_test/estates/fetch", dispatchEstates);
};

//initialState
const initialState = {
    estates: [],
};

//reducer
const estatesReducer = (state=initialState, action) => {
    switch(action.type) {
        case SET_ESTATES: return {
            estates: action.estates,
        };

        default: return state;
    };
};

//selectors
export const estatesSelector = (state) => (state.estates.estates);

export default estatesReducer;
