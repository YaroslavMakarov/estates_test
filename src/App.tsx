import React, { useEffect, Dispatch } from 'react';
import { Container } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import {Switch, Route } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import EstatesList from './components/EstatesList';
import EstateInfo from './components/EstatesInfo';

import { AppWrapper } from './styled/AppSteles';
import { AllEstatesAction, fetchPosts } from './redux/estates';


const App = () => {
  const estatesDispatch = useDispatch<Dispatch<AllEstatesAction>>();

  useEffect(() => {
    estatesDispatch(fetchPosts());
  }, []);

  return (
    <AppWrapper>
      <Container className="
          d-flex
          align-items-center
          flex-column
        "
      >
        <Switch>
          <Route
            path="/"
            exact
            component={EstatesList}
          />
          <Route
            path="/info/:id"
            exact
            component={EstateInfo}
          />
        </Switch>
      </Container>
    </AppWrapper>
  );
};

export default App;
