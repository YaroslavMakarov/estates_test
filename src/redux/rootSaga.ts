import { all } from 'redux-saga/effects';

import { dispatchEstatesWatcher } from './estates';

export function* rootSaga() {

    yield all([
        dispatchEstatesWatcher(),
    ])
}