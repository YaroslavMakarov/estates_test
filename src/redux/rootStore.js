import { combineReducers, createStore, applyMiddleware } from "redux";
import createSagaMiddleware from "redux-saga";

import estatesReducer from "./estates";
import { rootSaga } from "./rootSaga";

const rootReucer = combineReducers({
    estates: estatesReducer,
});

const reduxSagas = createSagaMiddleware();

const store = createStore(rootReucer, applyMiddleware(reduxSagas));

reduxSagas.run(rootSaga);

export default store;