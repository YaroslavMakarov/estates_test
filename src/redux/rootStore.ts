import { combineReducers, createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';

import { loadState, saveState } from '../helpers/localStorage';
import estatesReducer, { InitialEstateState } from './estates';
import { rootSaga } from './rootSaga';

export const rootReucer = combineReducers({
    estates: estatesReducer,
});

const persistedState = loadState();

const reduxSagas = createSagaMiddleware();

const store = createStore(rootReucer, persistedState, applyMiddleware(reduxSagas));

store.subscribe(() => {
    saveState(store.getState());
});

reduxSagas.run(rootSaga);

//root state type
export type RootState = {
    estates: InitialEstateState;
};

export default store;
