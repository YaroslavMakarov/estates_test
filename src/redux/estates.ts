import { put, takeEvery,
         call } from 'redux-saga/effects';
import { Action } from 'redux';         

import { getEstates } from '../helpers/api';
import { RootState } from './rootStore';       

//action types
export const FETCH_ESTATES = 'estates_test/estates/fetch';
const SET_ESTATES = 'estates_test/estates/set';

//action creator types
type FetchEstates = Action<typeof FETCH_ESTATES>;
type SetEstates = Action<typeof SET_ESTATES> & {
    estates: Array<Estate>;
}

//action creators
export const fetchPosts = (): FetchEstates => ({
    type: FETCH_ESTATES,
});
const setEstates = (estates: Array<Estate>): SetEstates => ({
    type: SET_ESTATES,
    estates,
});

//sagas
function* dispatchEstates() {
    const estates = yield call(getEstates);

    yield put(setEstates(estates));
};

export function* dispatchEstatesWatcher() {

    yield takeEvery('estates_test/estates/fetch', dispatchEstates);
};

//initialState Type
export type InitialEstateState = {
    estates: Array<Estate>
};

//initialState
const initialState: InitialEstateState = {
    estates: [],
};

//all action types
export type AllEstatesAction = FetchEstates | SetEstates;

//reducer
const estatesReducer = (state=initialState, action: AllEstatesAction) => {
    switch(action.type) {
        case SET_ESTATES: return {
            estates: action.estates,
        };

        default: return state;
    };
};

//selectors
export const estatesSelector = (state: RootState) => (state.estates.estates);

export default estatesReducer;
