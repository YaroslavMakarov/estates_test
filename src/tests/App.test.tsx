import React, { Dispatch } from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { Action } from 'redux';
import * as ReactRedux from 'react-redux';
import configureStore from 'redux-mock-store';

import App from '../App';
import { estatesMock } from '../mocks/estatesMock';
import { FETCH_ESTATES } from '../redux/estates';


const mockStore = configureStore();

describe('App component', () => {
    const store = mockStore(estatesMock);
    let wrapper: RenderResult;
    let useDispatch: jest.SpyInstance<Dispatch<Action<any>>, []>;
    const mockActionCreator = jest.fn();
    const mockDispatch = () => {
        useDispatch.mockReturnValue(mockActionCreator);
    }
    useDispatch = jest.spyOn(ReactRedux, 'useDispatch');

    beforeEach(() => {
        mockDispatch();
        wrapper = render(    
            <Provider store={store}>
                <MemoryRouter>
                    <App />
                </MemoryRouter>    
            </Provider>
        );
    });

    it('get snapshot App component', () => {
        expect(wrapper.baseElement).toMatchSnapshot();
    });

    it('shoud mock dispatch', () => {
        expect(mockActionCreator).toHaveBeenCalledWith({type: FETCH_ESTATES});
    });
});

