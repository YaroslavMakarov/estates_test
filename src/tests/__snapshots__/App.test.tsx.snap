// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component get snapshot App component 1`] = `
<body>
  <div>
    <div
      class="sc-jSgupP gQRInk"
    >
      <div
        class=" d-flex align-items-center flex-column  container"
      >
        <div
          class="sc-hKgILt gYRjSP"
        >
          <a
            class="sc-bdfBwQ fhkZDT"
            data-testid="estate"
            href="/info/1"
          >
            <img
              alt="main img"
              class="sc-gsTCUz esHetz"
              src="https://cp12.nevsepic.com.ua/62/1353962605-0435309-www.nevsepic.com.ua.jpg"
            />
            <div
              class="sc-dlfnbm hMEdSQ"
            >
              Name: Shevchenko estate
            </div>
            <div
              class="sc-dlfnbm hMEdSQ"
            >
              Price: 30000$
            </div>
            <div
              class="sc-dlfnbm hMEdSQ"
            >
              Address: Melnychna, 12
            </div>
          </a>
          <a
            class="sc-bdfBwQ fhkZDT"
            data-testid="estate"
            href="/info/2"
          >
            <img
              alt="main img"
              class="sc-gsTCUz esHetz"
              src="https://biz.liga.net/images/general/2018/11/15/thumbnail-tw-20181115111659-8474.jpg?v=1542279867"
            />
            <div
              class="sc-dlfnbm hMEdSQ"
            >
              Name: Yanukovich estate
            </div>
            <div
              class="sc-dlfnbm hMEdSQ"
            >
              Price: 3000000$
            </div>
            <div
              class="sc-dlfnbm hMEdSQ"
            >
              Address: Prisident, 4
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</body>
`;
